New record and file (#30)
skuid/SamplePages: Salesforce_Advanced/Create_Record_and_File/README.md

# Create New Record and File in One Step. 

> Salesforce content documents and attachments have to be connected to existing records.  This means that you cannot create a new record, and upload a file for that record in the same UI step. Typically Skuid builders have used a 2 step wizard to handle this,  enabling the data entry for a new record, then forcing the user to save that record before they can upload a file or image to it. This is functional,  but not optimal. 

Reviewer: ```suggestion
Salesforce content documents and attachments have to be connected to **existing records.** This means that you cannot create a new record and upload a file for that record in the same UI step. 

Typically Skuid builders have used a 2 step wizard to handle this, with step one being data entry for a new record, then forcing the user to save that record before they can upload a file or image to it in step two. This is functional, but not optimal. 
```

Reviewer: Not essential, but I think this makes it a little more clear.

> This sample page shows how to create a case, and upload a file to that case in the same step. Though this method seems like a bit of hack - its actually used behind the scenes by Salesforce in some of their standard UI. 

Reviewer: ```suggestion
This sample page shows how to create a case and upload a file to that case in the same step. Though this method seems like a bit of hack, it's actually used behind the scenes by Salesforce in some of their standard UI. 
```
Some grammar things.

> The process is as follows. 

Reviewer: ```suggestion
The process is as follows:
```

- File is initially uploaded as a content document linked to the `current user` record. (Everyone has a saved user record.) 
- After the new record metadata is completed the save button does the following: 
    - Saves the new record. 
    - Creates a new `content link` record connecting the file to the new record that has now been saved. 
    - Saves the content link record. 

Curiously there is always a content document link record created to the user record. You can't delete it. 

> We also include some Javscript and an Action Sequence that deletes the uploaded file if the page is abandoned before the new record is saved.  This keeps things clean. 

Reviewer: ```suggestion
We also include some JavaScript and an action sequence that deletes the uploaded file if the page is abandoned before the new record is saved.  This keeps things clean. 
```
